### PR TITLE
doc: Reformatted Storage Documentation to Organize Keys by Storage Type

### PR DESCRIPTION
**Pull request by jeff2814**
Signed-off-by: Jeffery Wu <jeff2814@hotmail.com>

### CODE REVIEW

**Comment by jeff2814** [comment]
fixes #9239

File: doc/storage.md [hunk 1]
```diff
@@ -2,65 +2,20 @@
...
-volume.block.filesystem         | string    | block based driver (lvm)          | ext4                       | Filesystem to use for new volumes
-volume.block.mount\_options     | string    | block based driver (lvm)          | discard                    | Mount options for block devices
-volume.size                     | string    | appropriate driver                | unlimited (10GiB for block)| Default volume size
-volume.zfs.remove\_snapshots    | bool      | zfs driver                        | false                      | Remove snapshots as needed
-volume.zfs.use\_refquota        | bool      | zfs driver                        | false                      | Use refquota instead of quota for space.
-zfs.clone\_copy                 | string    | zfs driver                        | true                       | Whether to use ZFS lightweight clones rather than full dataset copies (boolean) or "rebase" to copy based on the initial image.
-zfs.pool\_name                  | string    | zfs driver                        | name of the pool           | Name of the zpool
+- [directory](#directory)
+- [CEPH](#ceph)
```
Comment (stgraber) [inline doc/storage.md:8]: Missing `cephfs` which is another standalone storage driver

File: doc/storage.md [hunk 2]
```diff
@@ -2,65 +2,20 @@
...
-volume.block.filesystem         | string    | block based driver (lvm)          | ext4                       | Filesystem to use for new volumes
-volume.block.mount\_options     | string    | block based driver (lvm)          | discard                    | Mount options for block devices
-volume.size                     | string    | appropriate driver                | unlimited (10GiB for block)| Default volume size
-volume.zfs.remove\_snapshots    | bool      | zfs driver                        | false                      | Remove snapshots as needed
-volume.zfs.use\_refquota        | bool      | zfs driver                        | false                      | Use refquota instead of quota for space.
-zfs.clone\_copy                 | string    | zfs driver                        | true                       | Whether to use ZFS lightweight clones rather than full dataset copies (boolean) or "rebase" to copy based on the initial image.
-zfs.pool\_name                  | string    | zfs driver                        | name of the pool           | Name of the zpool
+- [directory](#directory)
```
Comment (stgraber) [inline doc/storage.md:7]: I'd call it `dir` just so the name matches the driver name

File: doc/storage.md [hunk 3]
```diff
@@ -221,6 +176,28 @@ This also means that access to cached data will not be affected by the limit.
...
 
+#### Storage pool configuration
+Key                             | Type      | Condition                         | Default                    | Description
+:--                             | :---      | :--------                         | :------                    | :----------
+source                          | string    | -                                 | -                          | Path to block device or loop file or filesystem entry
+rsync.bwlimit                   | string    | -                                 | 0 (no limit)               | Specifies the upper limit to be placed on the socket I/O whenever rsync has to be used to transfer storage entities.
+rsync.compression               | bool      | appropriate driver                | true                       | Whether to use compression while migrating storage pools.
+volatile.initial\_source        | string    | -                                 | -                          | Records the actual source passed during creating (e.g. /dev/sdb).
+volatile.pool.pristine          | string    | -                                 | true                       | Whether the pool has been empty on creation time.
```
Comment (stgraber) [inline doc/storage.md:186]: Those two don't apply to `dir` so should be dropped.

File: doc/storage.md [hunk 4]
```diff
@@ -221,6 +176,28 @@ This also means that access to cached data will not be affected by the limit.
...
+volume.size                     | string    | appropriate driver                | unlimited (10GiB for block)| Default volume size
+
+#### Storage volume configuration
+Key                     | Type      | Condition                 | Default                               | Description
+:--                     | :---      | :--------                 | :------                               | :----------
+size                    | string    | appropriate driver        | same as volume.size                   | Size of the storage volume
+block.filesystem        | string    | block based driver        | same as volume.block.filesystem       | Filesystem of the storage volume
+block.mount\_options    | string    | block based driver        | same as volume.block.mount\_options   | Mount options for block devices
```
Comment (stgraber) [inline doc/storage.md:194]: Those two don't apply to the `dir` driver and should be dropped.

File: doc/storage.md [hunk 5]
```diff
@@ -258,6 +235,39 @@ lxc storage create pool2 dir source=/data/lxd
...
+ceph.osd.data\_pool\_name       | string    | ceph driver                       | -                          | Name of the osd data pool.
+ceph.rbd.clone\_copy            | string    | ceph driver                       | true                       | Whether to use RBD lightweight clones rather than full dataset copies.
+ceph.rbd.features               | string    | ceph driver                       | layering                   | Comma separate list of RBD features to enable on the volumes.
+ceph.user.name                  | string    | ceph driver                       | admin                      | The ceph user to use when creating storage pools and volumes.
+cephfs.cluster\_name            | string    | cephfs driver                     | ceph                       | Name of the ceph cluster in which to create new storage pools.
+cephfs.path                     | string    | cephfs driver                     | /                          | The base path for the CEPHFS mount
+cephfs.user.name                | string    | cephfs driver                     | admin                      | The ceph user to use when creating storage pools and volumes.
```
Comment (stgraber) [inline doc/storage.md:257]: The `cephfs` ones only apply to the cephfs driver so should be moved to that one.

File: doc/storage.md [hunk 6]
```diff
@@ -258,6 +235,39 @@ lxc storage create pool2 dir source=/data/lxd
   storage entities is not recommended by Ceph upstream. You may see unexpected
   and erratic failures which are unrelated to LXD itself.
 
+#### Storage pool configuration
+Key                             | Type      | Condition                         | Default                    | Description
+:--                             | :---      | :--------                         | :------                    | :----------
+source                          | string    | -                                 | -                          | Path to block device or loop file or filesystem entry
+rsync.bwlimit                   | string    | -                                 | 0 (no limit)               | Specifies the upper limit to be placed on the socket I/O whenever rsync has to be used to transfer storage entities.
+rsync.compression               | bool      | appropriate driver                | true                       | Whether to use compression while migrating storage pools.
+volatile.initial\_source        | string    | -                                 | -                          | Records the actual source passed during creating (e.g. /dev/sdb).
```
Comment (stgraber) [inline doc/storage.md:244]: Not use for the ceph driver so should be dropped.

File: doc/storage.md [hunk 7]
```diff
@@ -300,6 +310,29 @@ lxc storage create pool1 ceph source=my-already-existing-osd
...
+volatile.pool.pristine          | string    | -                                 | true                       | Whether the pool has been empty on creation time.
```
Comment (stgraber) [inline doc/storage.md:320]: Not used by the btrfs driver, should be dropped

File: doc/storage.md [hunk 8]
```diff
@@ -300,6 +310,29 @@ lxc storage create pool1 ceph source=my-already-existing-osd
...
+volume.size                     | string    | appropriate driver                | unlimited (10GiB for block)| Default volume size
+btrfs.mount\_options            | string    | btrfs driver                      | user\_subvol\_rm\_allowed  | Mount options for block devices
+
+#### Storage volume configuration
+Key                     | Type      | Condition                 | Default                               | Description
+:--                     | :---      | :--------                 | :------                               | :----------
+size                    | string    | appropriate driver        | same as volume.size                   | Size of the storage volume
+block.filesystem        | string    | block based driver        | same as volume.block.filesystem       | Filesystem of the storage volume
+block.mount\_options    | string    | block based driver        | same as volume.block.mount\_options   | Mount options for block devices
```
Comment (stgraber) [inline doc/storage.md:329]: Not relevant to the btrfs driver should be dropped.

File: doc/storage.md [hunk 9]
```diff
@@ -350,6 +383,38 @@ sudo btrfs filesystem resize max /var/lib/lxd/storage-pools/<POOL>/
...
    LXD.
 
+#### Storage pool configuration
+Key                             | Type      | Condition                         | Default                    | Description
+:--                             | :---      | :--------                         | :------                    | :----------
+source                          | string    | -                                 | -                          | Path to block device or loop file or filesystem entry
+rsync.bwlimit                   | string    | -                                 | 0 (no limit)               | Specifies the upper limit to be placed on the socket I/O whenever rsync has to be used to transfer storage entities.
+rsync.compression               | bool      | appropriate driver                | true                       | Whether to use compression while migrating storage pools.
+volatile.initial\_source        | string    | -                                 | -                          | Records the actual source passed during creating (e.g. /dev/sdb).
+volatile.pool.pristine          | string    | -                                 | true                       | Whether the pool has been empty on creation time.
```
Comment (stgraber) [inline doc/storage.md:393]: Not relevant to the LVM driver, should be dropped.

File: doc/storage.md [hunk 10]
```diff
@@ -350,6 +383,38 @@ sudo btrfs filesystem resize max /var/lib/lxd/storage-pools/<POOL>/
...
+lvm.use\_thinpool               | bool      | lvm driver                        | true                       | Whether the storage pool uses a thinpool for logical volumes.
+lvm.vg\_name                    | string    | lvm driver                        | name of the pool           | Name of the volume group to create.
+lvm.vg.force\_reuse             | bool      | lvm driver                        | false                      | Force using an existing non-empty volume group.
+volume.lvm.stripes              | string    | lvm driver                        | -                          | Number of stripes to use for new volumes (or thin pool volume).
+volume.lvm.stripes.size         | string    | lvm driver                        | -                          | Size of stripes to use (at least 4096 bytes and multiple of 512bytes).
+volume.block.filesystem         | string    | block based driver (lvm)          | ext4                       | Filesystem to use for new volumes
+volume.block.mount\_options     | string    | block based driver (lvm)          | discard                    | Mount options for block devices
```
Comment (stgraber) [inline doc/storage.md:402]: So there is an inconsistency with those across the different drivers.
Basically you can set `volume.KEY` on the pool for any valid volume key defined below for the given driver.

So we either need to make sure to duplicate the entire set of volume keys into the pool config with a `volume.` prefix, or we should just drop them from all pool configs and make sure we have a dedicated section in the doc which explains the pool to volume relationship and that you can use `volume.KEY` on the pool for any valid volume key so you can set its default within that pool.

File: doc/storage.md [hunk 11]
```diff
@@ -423,6 +488,35 @@ lxc storage create pool1 lvm source=/dev/sdX lvm.vg_name=my-pool
...
 
+#### Storage pool configuration
+Key                             | Type      | Condition                         | Default                    | Description
+:--                             | :---      | :--------                         | :------                    | :----------
+source                          | string    | -                                 | -                          | Path to block device or loop file or filesystem entry
+rsync.bwlimit                   | string    | -                                 | 0 (no limit)               | Specifies the upper limit to be placed on the socket I/O whenever rsync has to be used to transfer storage entities.
+rsync.compression               | bool      | appropriate driver                | true                       | Whether to use compression while migrating storage pools.
+volatile.initial\_source        | string    | -                                 | -                          | Records the actual source passed during creating (e.g. /dev/sdb).
+volatile.pool.pristine          | string    | -                                 | true                       | Whether the pool has been empty on creation time.
```
Comment (stgraber) [inline doc/storage.md:498]: Not relevant to the zfs driver.

File: doc/storage.md [hunk 12]
```diff
@@ -423,6 +488,35 @@ lxc storage create pool1 lvm source=/dev/sdX lvm.vg_name=my-pool
...
+zfs.clone\_copy                 | string    | zfs driver                        | true                       | Whether to use ZFS lightweight clones rather than full dataset copies (boolean) or "rebase" to copy based on the initial image.
+zfs.pool\_name                  | string    | zfs driver                        | name of the pool           | Name of the zpool
+
+#### Storage volume configuration
+Key                     | Type      | Condition                 | Default                               | Description
+:--                     | :---      | :--------                 | :------                               | :----------
+size                    | string    | appropriate driver        | same as volume.size                   | Size of the storage volume
+block.filesystem        | string    | block based driver        | same as volume.block.filesystem       | Filesystem of the storage volume
+block.mount\_options    | string    | block based driver        | same as volume.block.mount\_options   | Mount options for block devices
```
Comment (stgraber) [inline doc/storage.md:511]: Doesn't apply to zfs.

**Comment by stgraber** [comment]
Alright, I think I went through all of it.
The summary is:
 - There are a bunch of keys that shouldn't be included for a variety of storage drivers. I think I added notes for all of those.
 - We need a separate section for `cephfs` distinct from `ceph`.
 - I think I'd prefer the storage drivers to match their LXD name in the list at the top (so, `ceph`, `cephfs`, `dir`, `btrfs`, `lvm` and `zfs`)
 - Need to sort out the handling of `volume.XYZ` keys, most likely by not having those included in every table but instead having a clear explanation that anything that's valid on a volume for a particular pool can be set as `volume.KEY=VALUE` on the pool and will then be used as a default for new volumes.

**Comment by tomponline** [comment]
> I think I'd prefer the storage drivers to match their LXD name in the list at the top (so, `ceph`, `cephfs`, `dir`, `btrfs`, `lvm` and `zfs`)

This would also be consistent with instance device type headings.

File: doc/storage.md [hunk 13]
```diff
@@ -2,71 +2,33 @@
...
-volume.block.filesystem         | string    | block based driver (lvm)          | ext4                       | Filesystem to use for new volumes
-volume.block.mount\_options     | string    | block based driver (lvm)          | discard                    | Mount options for block devices
-volume.size                     | string    | appropriate driver                | unlimited (10GiB for block)| Default volume size
-volume.zfs.remove\_snapshots    | bool      | zfs driver                        | false                      | Remove snapshots as needed
-volume.zfs.use\_refquota        | bool      | zfs driver                        | false                      | Use refquota instead of quota for space.
-zfs.clone\_copy                 | string    | zfs driver                        | true                       | Whether to use ZFS lightweight clones rather than full dataset copies (boolean) or "rebase" to copy based on the initial image.
-zfs.pool\_name                  | string    | zfs driver                        | name of the pool           | Name of the zpool
+- [dir](#directory)
```
Comment (stgraber) [inline doc/storage.md:7]: I'd also rename the section from directory to dir

File: doc/storage.md [hunk 14]
```diff
@@ -258,6 +237,31 @@ lxc storage create pool2 dir source=/data/lxd
   storage entities is not recommended by Ceph upstream. You may see unexpected
   and erratic failures which are unrelated to LXD itself.
 
+#### Storage pool configuration
+Key                             | Type      | Condition                         | Default                    | Description
+:--                             | :---      | :--------                         | :------                    | :----------
+volatile.pool.pristine          | string    | -                                 | true                       | Whether the pool has been empty on creation time.
```
Comment (stgraber) [inline doc/storage.md:243]: The config keys should be sorted alphabetically so this one should go to the end of the list, same for any other list of keys which isn't currently in alphabetical order.

File: doc/storage.md [hunk 15]
```diff
@@ -258,6 +237,31 @@ lxc storage create pool2 dir source=/data/lxd
...
+ceph.osd.force\_reuse           | bool      | ceph driver                       | false                      | Force using an osd storage pool that is already in use by another LXD instance.
+ceph.osd.pg\_num                | string    | ceph driver                       | 32                         | Number of placement groups for the osd storage pool.
+ceph.osd.pool\_name             | string    | ceph driver                       | name of the pool           | Name of the osd storage pool.
+ceph.osd.data\_pool\_name       | string    | ceph driver                       | -                          | Name of the osd data pool.
+ceph.rbd.clone\_copy            | string    | ceph driver                       | true                       | Whether to use RBD lightweight clones rather than full dataset copies.
+ceph.rbd.features               | string    | ceph driver                       | layering                   | Comma separate list of RBD features to enable on the volumes.
+ceph.user.name                  | string    | ceph driver                       | admin                      | The ceph user to use when creating storage pools and volumes.
```
Comment (stgraber) [inline doc/storage.md:251]: The `condition` column can be dropped from all storage pool configs I believe as they're now irrelevant.
The ones for the volumes still appear relevant though so we should keep those.

**Comment by stgraber** [comment]
Looking good. I think we just need:
 - Sort all config keys alphabetically
 - Drop the now redundant `condition` column from the list of pool config keys
 - Rename the `directory` section to `dir`

**Comment by stgraber** [comment]
Looks good, just a small issue with cephfs as it does support some of the ceph keys, but you couldn't have known that. I'll push an extra commit on top of this to sort that out and then merge it.

**Comment by stgraber** [comment]
jenkins: test this please

**Pull request merged**